src/OpenRiaServices.DomainServices.Hosting/Framework/Services/MessageEncoders/PoxBinaryMessageEncodingBindingElement.cs

```diff
@@ -255,15 +256,207 @@ public override ArraySegment<byte> WriteMessage(Message message, int maxMessageS
 
                 message.Properties.Encoder = this;
 
-                using (MemoryStream ms = new MemoryStream())
-                using (XmlDictionaryWriter writer = XmlDictionaryWriter.CreateBinaryWriter(ms))
+                /// PERF: 
+                /// For further imprioved perf we should look into adopting a similar behaviour as for the 
+                /// in binary encoding which performs *size prediction* 
+                /// https://referencesource.microsoft.com/#System.ServiceModel/System/ServiceModel/Channels/BufferedMessageWriter.cs,ed72c7ce79a15637
+                /// that should allow us to skip memory copies and further improve performance.
+                /// 
+                /// We should be able to pool both the stream and the binary writer togheter with size data
+                using (var stream = new BufferManagerStream(bufferManager, messageOffset, minAllocationSize: 2 * 1024, maxAllocationSize: maxMessageSize))
```

Comment:
the current min and max size are arbitrary

One could consider setting a fixed max size at 64k since that is just below the LOH threashold.
But since the memory should be pooled anyway the arrays should probably end up in Gen2 when everything works as expected